### PR TITLE
DAOS-4181 sec: Add access capabilities for pools/conts

### PR DESCRIPTION
**Pull request by kjacque**
This patch adds new functions to get access capabilities
for pools and containers based on their ACL, ownership,
the caller's user credential, and request flags.

- Add new bitmasks for pool and container security
  capabilities.
- Replace the previous pool access check with new method
  ds_sec_pool_get_capabilities.
- Save the pool security capabilities in the pool handle
  on the server side. They are not exposed to the client.
- Rename internal usages of the old "capas" to "flags" to
  better describe their function and differentiate from
  the security capas.
- Add ds_sec_cont_get_capabilities to allow similar
  functionality for containers in future, although it is
  not yet used.
- Add ds_sec_pool_can_connect to verify pool connections
  based on access capabilities.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

Requesting exception as only test failures were due to a post provisioning bug in pipeline-lib https://github.com/daos-stack/pipeline-lib/commit/62633799c9795169a5806390ef443bfbcb238d65

### CODE REVIEW

File: src/security/srv_acl.c [hunk 1]
```diff
@@ -409,60 +379,142 @@ check_access_for_groups(struct daos_acl *acl,
...
 }
 
 static int
-check_authsys_permissions(struct daos_acl *acl,
-			  struct pool_owner *ownership,
-			  Auth__Sys *authsys, uint64_t capas)
+get_authsys_capas(struct daos_acl *acl,
+		  struct ownership *ownership,
+		  Auth__Sys *authsys,
+		  uint64_t(*convert_perms)(uint64_t),
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/security/srv_acl.c]: Fixed

File: src/security/srv_acl.c [hunk 2]
```diff
@@ -473,34 +525,138 @@ ds_sec_check_pool_access(struct daos_acl *acl, struct pool_owner *ownership,
...
+	return capas;
+}
+
+static void
+filter_cont_capas_based_on_flags(uint64_t flags, uint64_t *capas)
+{
+	if (flags & DAOS_COO_RO)
+		*capas &= CONT_CAPAS_RO_MASK;
+	else if (!(*capas & CONT_CAPAS_RO_MASK) ||
+	         !(*capas & ~CONT_CAPAS_RO_MASK))
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/security/srv_acl.c]: Fixed

File: src/security/srv_acl.c [hunk 3]
```diff
@@ -409,60 +379,142 @@ check_access_for_groups(struct daos_acl *acl,
...
+	if (daos_acl_principal_is_valid(ownership->user) &&
+	    daos_acl_principal_is_valid(ownership->group))
+		return true;
+
+	return false;
+}
+
+static int
+get_sec_capas_for_token(Auth__Token *token, struct ownership *ownership,
+		        struct daos_acl *acl,
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/security/srv_acl.c]: Fixed

File: src/security/srv_acl.c [hunk 4]
```diff
@@ -409,60 +379,142 @@ check_access_for_groups(struct daos_acl *acl,
...
+		        uint64_t(*convert_perms)(uint64_t), uint64_t *capas)
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/security/srv_acl.c]: Fixed

File: src/security/srv_acl.c [hunk 5]
```diff
@@ -409,60 +379,142 @@ check_access_for_groups(struct daos_acl *acl,
...
+	return 0;
+}
+
+static void
+filter_pool_capas_based_on_flags(uint64_t flags, uint64_t *capas)
+{
+	if (flags & DAOS_PC_RO)
+		*capas &= POOL_CAPAS_RO_MASK;
+	else if (!(*capas & POOL_CAPAS_RO_MASK) ||
+	         !(*capas & ~POOL_CAPAS_RO_MASK))
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/security/srv_acl.c]: Fixed

File: src/container/srv_container.c [hunk 6]
```diff
@@ -548,8 +548,8 @@ cont_create(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl,
 		DP_CONT(pool_hdl->sph_pool->sp_uuid, in->cci_op.ci_uuid), rpc);
 
 	/* Verify the pool handle capabilities. */
-	if (!(pool_hdl->sph_capas & DAOS_PC_RW) &&
-	    !(pool_hdl->sph_capas & DAOS_PC_EX))
+	if (!(pool_hdl->sph_flags & DAOS_PC_RW) &&
+	    !(pool_hdl->sph_flags & DAOS_PC_EX))
```
Comment (NiuYawei) [inline src/container/srv_container.c]: looks 'mode' is a better name.

**Comment by jolivier23** [comment]
There is a conflict to be resolved on src/pool/srv_pool.c

**Comment by kjacque** [comment]
One of the test failures is [DAOS-4025](https://jira.hpdd.intel.com/browse/DAOS-4025), rebuild 19 test timeout.
There is another failure in the rebuild 15 test, looks like [DAOS-4171](https://jira.hpdd.intel.com/browse/DAOS-4171).

**Pull request merged**